### PR TITLE
retire asset in Teamdynamix with host before_destroy hook

### PR DESCRIPTION
**Pull request by nipendar**
any failure in creating/deleing asset in TD will be displayed to the user via AR Errors

### CODE REVIEW

File: app/models/concerns/foreman_teamdynamix/host_extensions.rb [hunk 1]
```diff
@@ -16,7 +17,14 @@ def td_api
     def create_teamdynamix_asset
       asset = td_api.create_asset(self)
       self.teamdynamix_asset_id = asset['ID']
+    rescue StandardError => e
+      errors.add(:base, e.message)
```
Comment (edestecd) [inline app/models/concerns/foreman_teamdynamix/host_extensions.rb]: Please follow this example: https://github.com/theforeman/puppetdb_foreman/blob/master/app/models/puppetdb_foreman/host_extensions.rb#L22

Use the language translation method: _()
Also return false. Some things in rails expect this on failure.

File: app/models/concerns/foreman_teamdynamix/host_extensions.rb [hunk 2]
```diff
@@ -16,7 +17,14 @@ def td_api
...
     end
 
+    def retire_teamdynamix_asset
+      td_api.retire_asset(self.teamdynamix_asset_id) if self.teamdynamix_asset_id
+    rescue StandardError => e
+      errors.add(:base, e.message)
```
Comment (edestecd) [inline app/models/concerns/foreman_teamdynamix/host_extensions.rb]: Please follow this example: https://github.com/theforeman/puppetdb_foreman/blob/master/app/models/puppetdb_foreman/host_extensions.rb#L22

Use the language translation method: _()
Also return false. Some things in rails expect this on failure.

File: lib/teamdynamix_api.rb [hunk 3]
```diff
@@ -93,16 +98,21 @@ def valid_auth_token?(token)
     token.match(/^[a-zA-Z0-9\.\-\_]*$/)
   end
 
+  def retire_asset_payload(asset_id)
+    asset = get_asset(asset_id)
+    asset.merge!(API_CONFIG[:delete].stringify_keys)
```
Comment (edestecd) [inline lib/teamdynamix_api.rb]: Hmm, I think the yaml parser can be configured to stringify the keys or switch to hash with indeffernent access. Does the foreman config loader not do this? Did you actually have an issue here?

Comment (nipendar) [reply lib/teamdynamix_api.rb]: asset returned has stringified keys so in order to overwrite the StatusID I am stringifying that key along with any other attributes user may have configured to be updated on retire.

is there something wrong with using stringify_keys?

Comment (edestecd) [reply lib/teamdynamix_api.rb]: No, I'm just wondering if Foreman Core is already doing it or similar.
Did you inspect to make sure its not already handled?

**Pull request merged**